libensemble/generators.py

```diff
@@ -0,0 +1,300 @@
+                self.model = None
+
+            def ask(self, num_points):
+                return create_points(num_points, self.param)
+
+            def tell(self, results):
+                self.model = update_model(results, self.model)
+
+            def final_tell(self, results):
+                self.tell(results)
+                return list(self.model)
+
+
+        my_generator = MyGenerator(my_parameter=100)
+        gen_specs = GenSpecs(generator=my_generator, ...)
+    """
+
+    @abstractmethod
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the Generator object on the user-side. Constants, class-attributes,
+        and preparation goes here.
+
+        .. code-block:: python
+
+            my_generator = MyGenerator(my_parameter, batch_size=10)
+        """
+
+    @abstractmethod
+    def ask(self, num_points: Optional[int]) -> List[dict]:
+        """
+        Request the next set of points to evaluate.
+        """
+
+    def ask_updates(self) -> npt.NDArray:
+        """
+        Request any updates to previous points, e.g. minima discovered, points to cancel.
+        """
+
+    def tell(self, results: List[dict]) -> None:
+        """
+        Send the results of evaluations to the generator.
+        """
+
+    def final_tell(self, results: List[dict], *args, **kwargs) -> Optional[npt.NDArray]:
+        """
+        Send the last set of results to the generator, instruct it to cleanup, and
+        optionally retrieve an updated final state of evaluations. This is a separate
+        method to simplify the common pattern of noting internally if a
+        specific tell is the last. This will be called only once.
+        """
+
+
+class LibensembleGenerator(Generator):
+    """Internal implementation of Generator interface for use with libEnsemble, or for those who
+    prefer numpy arrays. ``ask/tell`` methods communicate lists of dictionaries, like the standard.
+    ``ask_np/tell_np`` methods communicate numpy arrays containing the same data.
+    """
+
+    @abstractmethod
+    def ask_numpy(self, num_points: Optional[int] = 0) -> npt.NDArray:
+        pass
+
+    @abstractmethod
+    def tell_numpy(self, results: npt.NDArray) -> None:
+        pass
+
+    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
+        """Request the next set of points to evaluate."""
+        return np_to_list_dicts(self.ask_numpy(num_points))
+
+    def tell(self, calc_in: List[dict]) -> None:
+        """Send the results of evaluations to the generator."""
+        self.tell_numpy(list_dicts_to_np(calc_in))
+
+
+class LibensembleGenThreadInterfacer(LibensembleGenerator):
+    """Implement ask/tell for traditionally written libEnsemble persistent generator functions.
+    Still requires a handful of libEnsemble-specific data-structures on initialization.
+    """
+
+    def __init__(
+        self, gen_specs: dict, History: npt.NDArray = [], persis_info: dict = {}, libE_info: dict = {}
```

Comment:
Here we have a different constructor signature to those functions in gen_classes.